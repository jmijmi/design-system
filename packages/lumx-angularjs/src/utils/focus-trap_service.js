import { TAB_KEY_CODE } from '@lumx/core/js/constants';

/////////////////////////////

function FocusTrapService() {
    'ngInject';

    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The active element where to trap the focus.
     *
     * @type {Element}
     */
    let _activeElement;

    /////////////////////////////
    //                         //
    //    Private functions    //
    //                         //
    /////////////////////////////

    /**
     * Handle key events on key down.
     *
     * @param {Event} evt The key event.
     */
    function _onKeyDown(evt) {
        if (evt.keyCode !== TAB_KEY_CODE) {
            return;
        }

        const focusableEls = _activeElement.find(
            'a[href]:not([tabindex="-1"]), button:not([tabindex="-1"]), textarea:not([tabindex="-1"]), input[type="text"]:not([tabindex="-1"]), input[type="radio"]:not([tabindex="-1"]), input[type="checkbox"]:not([tabindex="-1"]), [tabindex]:not([tabindex="-1"])',
        );

        const firstFocusableEl = focusableEls[0];
        const lastFocusableEl = focusableEls[focusableEls.length - 1];

        if (evt.shiftKey) {
            if (document.activeElement === firstFocusableEl) {
                lastFocusableEl.focus();
                evt.preventDefault();
            }
        } else if (document.activeElement === lastFocusableEl) {
            firstFocusableEl.focus();
            evt.preventDefault();
        }
    }

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Activate focus trap on given element.
     *
     * @param {element} el The element where to trap focus.
     */
    function activate(el) {
        _activeElement = el;
        _activeElement.on('keydown', _onKeyDown);
    }

    /**
     * Disable focus trap on given element.
     */
    function disable() {
        _activeElement.off('keydown', _onKeyDown);
        _activeElement = undefined;
    }

    /////////////////////////////

    service.activate = activate;
    service.disable = disable;
}

/////////////////////////////

angular.module('lumx.utils.focus-trap').service('LxFocusTrapService', FocusTrapService);

/////////////////////////////

export { FocusTrapService };
