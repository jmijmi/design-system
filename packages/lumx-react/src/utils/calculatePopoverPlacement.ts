import { Placement } from '@lumx/react/components/popover/Popover';

export const calculatePopoverPlacement = (
    placement: Placement,
    anchor: ClientRect,
    popover: ClientRect,
): {
    x: number;
    y: number;
} => {
    const splittedPlacement = placement.split('-');
    const verticalPlacement = splittedPlacement[0];
    const horizontalPlacement = splittedPlacement[1];
    const { top, left, width: widthAnchor, height: heightAnchor } = anchor;
    const { width: widthPopover, height: heightPopover }: ClientRect | DOMRect = popover;
    let x = 0;
    let y = 0;
    switch (verticalPlacement) {
        case 'top':
            y += top - heightPopover;

            switch (horizontalPlacement) {
                case 'start':
                    x += left;

                    break;
                case 'end':
                    x += left + widthAnchor - widthPopover;
                    break;

                default:
                    x += left + (widthAnchor - widthPopover) / 2;

                    break;
            }

            break;
        case 'right':
            x += left + widthAnchor;

            switch (horizontalPlacement) {
                case 'start':
                    y += top;

                    break;
                case 'end':
                    y += top + heightAnchor - heightPopover;

                    break;

                default:
                    y += top + (heightAnchor - heightPopover) / 2;

                    break;
            }

            break;
        case 'bottom':
            y += top + heightAnchor;

            switch (horizontalPlacement) {
                case 'start':
                    x += left;

                    break;
                case 'end':
                    x += left + widthAnchor - widthPopover;

                    break;

                default:
                    x += left + (widthAnchor - widthPopover) / 2;

                    break;
            }

            break;
        case 'left':
            x += left - widthPopover;

            switch (horizontalPlacement) {
                case 'start':
                    y += top;

                    break;
                case 'end':
                    y += top + heightAnchor - heightPopover;

                    break;

                default:
                    y += top + (heightAnchor - heightPopover) / 2;

                    break;
            }

            break;

        default:
            break;
    }

    return { x, y };
};
