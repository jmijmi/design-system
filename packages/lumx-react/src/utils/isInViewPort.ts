export type ViewPortVisibility = 'full' | 'partial';

/**
 * Test if an element is in the viewport.
 *
 * @param bounding Bounding of element to test.
 * @param visibility Visible only partially or fully.
 * @return Whether the element to test is in viewport.
 */
export const isInViewPort = (bounding: ClientRect | DOMRect, visibility: ViewPortVisibility): boolean => {
    if (visibility === 'partial') {
        return (
            bounding.top <= (window.innerHeight || document.documentElement.clientHeight) &&
            bounding.left <= (window.innerWidth || document.documentElement.clientWidth) &&
            bounding.right >= 0 &&
            bounding.bottom >= 0
        );
    }

    return (
        bounding.top >= 0 &&
        bounding.left >= 0 &&
        bounding.right <= (window.innerWidth || document.documentElement.clientWidth) &&
        bounding.bottom <= (window.innerHeight || document.documentElement.clientHeight)
    );
};
