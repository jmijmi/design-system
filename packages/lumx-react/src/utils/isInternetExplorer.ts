/**
 * Check if browser is IE
 * @return Browser is IE or not
 */
const isInternetExplorer = () => {
    const ua = window.navigator.userAgent;
    const msie = ua.indexOf('MSIE ');
    const isIEVersion = !!navigator.userAgent.match(/Trident.*rv\:11\./);
    return msie > 0 || isIEVersion;
};

export { isInternetExplorer };
