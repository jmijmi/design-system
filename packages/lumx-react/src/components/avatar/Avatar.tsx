import React, { CSSProperties, ReactNode } from 'react';

import classNames from 'classnames';

import { Size, Theme } from '@lumx/react';
import { COMPONENT_PREFIX } from '@lumx/react/constants';
import { GenericProps, handleBasicClasses, getRootClassName } from '@lumx/react/utils';

/**
 * Authorized size values.
 */
export type AvatarSize = Size.xs | Size.s | Size.m | Size.l | Size.xl | Size.xxl;

/**
 * Defines the props of the component.
 */
export interface AvatarProps extends GenericProps {
    /** Actions elements to be transcluded into the component */
    actions?: HTMLElement | ReactNode;
    /** Size. */
    size?: AvatarSize;
    /** Theme. */
    theme?: Theme;
    /** Avatar image */
    image: string;
}

/**
 * The display name of the component.
 */
const COMPONENT_NAME = `${COMPONENT_PREFIX}Avatar`;

/**
 * The default class name and classes prefix for this component.
 */
export const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * The default value of props.
 */
export const DEFAULT_PROPS: Partial<AvatarProps> = {
    actions: undefined,
    size: Size.m,
    theme: Theme.light,
};

/**
 * Simple component used to identify user.
 *
 * @return The component.
 */
export const Avatar: React.FC<AvatarProps> = ({
    actions = DEFAULT_PROPS.actions,
    className,
    size = DEFAULT_PROPS.size,
    theme = DEFAULT_PROPS.theme,
    image,
    ...props
}) => {
    const style: CSSProperties = {
        backgroundImage: `url(${image})`,
    };
    return (
        <div
            className={classNames(
                className,
                handleBasicClasses({ prefix: CLASSNAME, size, variant: 'rounded', theme }),
            )}
            {...props}
            style={style}
        >
            {actions && <div className={`${CLASSNAME}__actions`}>{actions}</div>}
        </div>
    );
};
Avatar.displayName = COMPONENT_NAME;
