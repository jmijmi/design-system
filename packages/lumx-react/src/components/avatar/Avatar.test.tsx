import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { CommonSetup, Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';

import { Avatar, AvatarProps, CLASSNAME } from './Avatar';

/**
 * Defines what the `setup` function will return.
 */
interface Setup extends CommonSetup {
    avatar: Wrapper;
    props: Partial<AvatarProps>;

    /**
     * The <div> element wrapper.
     */
    wrapper: Wrapper;
}

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param props  The props to use to override the default props of the component.
 * @param     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return      An object with the props, the component wrapper and some shortcut to some element inside of the
 *                       component.
 */
const setup = ({ ...propsOverrides }: Partial<AvatarProps> = {}, shallowRendering: boolean = true): Setup => {
    const props: AvatarProps = {
        image: 'http://i.pravatar.cc/40',
        ...propsOverrides,
    };

    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<Avatar {...props} />);

    return {
        avatar: wrapper.find(`div`),
        props,
        wrapper,
    };
};

describe(`<${Avatar.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render correctly', () => {
            const { wrapper, avatar } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(avatar).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'avatar', prop: 'avatar' }, { className: CLASSNAME });
});
