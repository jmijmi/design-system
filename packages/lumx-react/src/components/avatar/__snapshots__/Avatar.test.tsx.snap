// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-avatar lumx-avatar--size-m lumx-avatar--variant-rounded lumx-avatar--theme-light"
  style={
    Object {
      "backgroundImage": "url(http://i.pravatar.cc/40)",
    }
  }
/>
`;
